pep-0544.txt

```diff
@@ -0,0 +1,968 @@
+PEP: 544
+Title: Protocols
+Version: $Revision$
+Last-Modified: $Date$
+Author: Ivan Levkivskyi <levkivskyi@gmail.com>, Jukka Lehtosalo <jukka.lehtosalo@iki.fi>, Łukasz Langa <lukasz@langa.pl>
+Discussions-To: Python-Dev <python-dev@python.org>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 05-Mar-2017
+Python-Version: 3.7
+
+
+Abstract
+========
+
+Type hints introduced in PEP 484 can be used to specify type metadata
+for static type checkers and other third party tools. However, PEP 484
+only specifies the semantics of *nominal* subtyping. In this PEP we specify
+static and runtime semantics of protocol classes that will provide a support
+for *structural* subtyping (static duck typing).
+
+
+.. _rationale:
+
+Rationale and Goals
+===================
+
+Currently, PEP 484 and the ``typing`` module [typing]_ define abstract
+base classes for several common Python protocols such as ``Iterable`` and
+``Sized``. The problem with them is that a class has to be explicitly marked
+to support them, which is unpythonic and unlike what one would
+normally do in idiomatic dynamically typed Python code. For example,
+this conforms to PEP 484::
+
+  from typing import Sized, Iterable, Iterator
+
+  class Bucket(Sized, Iterable[int]):
+      ...
+      def __len__(self) -> int: ...
+      def __iter__(self) -> Iterator[int]: ...
+
+The same problem appears with user-defined ABCs: they must be explicitly
+subclassed or registered. This is particularly difficult to do with library
+types as the type objects may be hidden deep in the implementation
+of the library. Moreover, extensive use of ABCs might impose additional
+runtime costs.
```

Comment:
```python
class C(collections.abc.Iterable):
    def __iter__(self):
        return []
```
is 2x slower than
```python
class C:
    def __iter__(self):
        return []
```
The numbers for ``typing.Iterable`` are even worse, but I have some ideas on how to improve those. The 2x slowdown is still not much, so I use "extensive" and "might". If you think it is not necessary we could remove this altogether.